Lib/pathlib/__init__.py

```diff
@@ -272,11 +272,18 @@ def iterdir(self):
         The children are yielded in arbitrary order, and the
         special entries '.' and '..' are not included.
         """
-        return (self._make_child_relpath(name) for name in os.listdir(self))
+        with os.scandir(self) as entries:
+            return iter([self._make_child_entry(entry) for entry in entries])
 
     def _scandir(self):
         return os.scandir(self)
 
+    def _make_child_entry(self, entry):
+        # Transform an entry yielded from _scandir() into a path object.
+        path = self.with_segments(entry.path)
+        path._str = entry.name if str(self) == '.' else entry.path
+        return path
```

Comment:
Why not use `self / entry.name`?

Reply:
`self / entry.name` is around 4x slower for the glob test case:

```shell
$ ./python -m timeit -s "from pathlib import Path" "list(Path.cwd().glob('**/*', follow_symlinks=True))"
5 loops, best of 5: 70 msec per loop  # before
1 loop, best of 5: 251 msec per loop  # with `self / entry.name`
```

This is because it does not maintain a normalized string representation, and so we hit the slow `_parse_path()`, `_format_parsed_parts()` codepaths for every directory we visit.